File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        parent = list(range(n))
        size = [1]*n
        
        def find_parent(a):
            if parent[a] != a:
                parent[a] = find_parent(parent[a])
            
            return parent[a]
        
        def union(a, b):
            pa = find_parent(a)
            pb = find_parent(b)
            
            if pa == pb:
                return
            pbig, psmall = (pa, pb) if size[pa] > size[pb] else (pb, pa)
            size[pbig] += psmall
            parent[psmall] = pbig
        
        for i in range(n):
            for j in range(i+1, n):
                if isConnected[i][j]:
                    union(i, j)
        
        return len(set(find_parent(i) for i in range(n)))

File: 0547-number-of-provinces/0547-number-of-provinces.cpp
class Solution {
public:
    void dfs(int node,vector<vector<int>>& isConnected,vector<bool>& vis){
        vis[node] = true;
        for(int i=0;i<isConnected[node].size();i++){
            if(!vis[i] and isConnected[node][i])
                dfs(i,isConnected,vis);
        }
    }
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size(), ans = 0;
        vector<bool> visited(n,false);
        for(int i=0;i<n;i++){
            if(!visited[i]){
                ans++;
                dfs(i,isConnected,visited);
            }
        }
        return ans;
    }
};